### DISCUSSION TITLE
Simple FPS Multiplayer Example

### DISCUSSION THREAD

**Discussion by proc-gen** [genaray/Arch#139] [Show and tell]
I'm back with another test project, this time implementing multiplayer with MonoGame and Arch - https://github.com/proc-gen/FpsGame

In it's current state, you can play the game in single player, client hosted multiplayer, joining a multiplayer game, and running as a dedicated host. Multiplayer works as long as the computers are on the same network, 

For those interested in how I handled the serialization and syncing between client and server, I would start with this part of the project - https://github.com/proc-gen/FpsGame/tree/master/FpsGame.Common/Serialization. I created `SerializableWorld` and `SerializableEntity` to handle moving the data back and forth. The supporting functions I needed to handle creating entities from a dynamic set of components are in https://github.com/proc-gen/FpsGame/blob/master/FpsGame.Common/Ecs/WorldExtensions.cs. 

As of right now I haven't delved too much into the actual game-play with more of the focus being on multiplayer and some other quality of life features. It is cross-platform, with the only caveat so far being that ARM-based Macs can't load textures.

I'll stop here before I go on too much :)

**Comment by genaray** [comment]
Sounds great! :) is it still WIP? And are there ingame screenshots? 

I'll look at the code later, thats gonna be interesting ^^

**Reply by proc-gen** [reply]
It's very much a WIP right now. The players are still capsules that seemingly float around without shadows and there's no projectiles or anything of the sort yet 😅 

I want to get the level generation stuff away from being hard-coded and then I plan to get back to more game-play type stuff.

**Reply by proc-gen** [reply]
![image](https://github.com/genaray/Arch/assets/65076703/fddc5cfa-4d97-435b-89dc-20bbe8106ba2)

**Reply by genaray** [reply]
Still looks great tho! :) did not knew that monogame is capable of 3D rendering. Excited to see how this one develops!

**Reply by proc-gen** [reply]
Thanks! I was running through some stuff and found a networking library I might be able to use to replace my TCP connection usage with UDP - https://github.com/lidgren/lidgren-network-gen3. Mainly, I'd want it for the port forwarding because right now my code requires being on the same network.

3D with MonoGame isn't all that bad unless you want to be able to build the project on an ARM based system. I gave a presentation where I knew everyone would have M1 based Macs, so I was trying everything I could to get the textures to load. Going forward, I may just add a blurb that it needs to be built on another system so the content gets compiled correctly until MonoGame fully supports M1's.

**Reply by genaray** [reply]
I can also recommend litenetlib. That one also uses udp instead of tcp and is great to use :) however lidgren is also a good alternative ^^ still pretty cool what monogame is capable of. I wonder how much it will develop with the 300k funding from re-logic :o
